lib/json/common.rb

```diff
@@ -143,7 +143,23 @@ class CircularDatastructure < NestingError; end
   # :startdoc:
 
   # This exception is raised if a generator or unparser error occurs.
-  class GeneratorError < JSONError; end
+  class GeneratorError < JSONError
+    attr_reader :invalid_object
+
+    def initialize(message, invalid_object = nil)
+      super(message)
+      @invalid_object = invalid_object
+    end
+
+    def detailed_message(...)
+      if @invalid_object.nil?
+        super
+      else
+        "#{super}\nInvalid object: #{@invalid_object.inspect}"
```

Comment:
The only quibble I've run into when pulling this into Sidekiq is the newline here. It's injecting formatting into the message, typically if I want to log an error message I don't expect it to take multiple lines in the log output. WDYT?

Reply:
Of course we expect newlines with backtraces... Could exceptions support a hash of context values to be logged with the error, where `invalid_object` would already be there? This is pretty standard stuff for structured logging, e.g. [Golang](https://pkg.go.dev/log/slog#Logger.With).

```ruby
rescue JSON::GeneratorError => ex
  logger.info { ex.with_context(some: value, another: value).detailed_message }
```

I guess the bigger question would be if ruby-core wants to move toward more formalized structured logging, by providing these types of APIs.